### PR TITLE
Streamk v0.3

### PR DESCRIPTION
**Pull request by xiaohuguo2023**
streamk gemm script v0.3

- new persistent gemm kernel
- gemm benchmark tool using nearest neighbor approach.
- new grid_model from Ryan
- updated kernel interfaces and tuning scripts
- reimplement stream-k atomic kernel which has no spills for large tile size
- add bf16 support

### CODE REVIEW

File: python/perf-kernels/streamk/utils/gemm_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            b.stride(1),
+            c.stride(0),
+            c.stride(1),
+            stride_bias,
+            BLOCK_SIZE_M=BLK_M,
+            BLOCK_SIZE_N=BLK_N,
+            BLOCK_SIZE_K=BLK_K,
+            GROUP_SIZE_M=gsize_m,
+            NUM_SMS=total_programs_streamk,
+            #            STREAMK_TILES=total_tiles_streamk,
```
Comment (neoblizz) [inline python/perf-kernels/streamk/utils/gemm_wrapper.py]: Shouldn't this be uncommented? Since all kernel signatures have `STREAMK_TILES` as a parameter?

File: python/perf-kernels/streamk/README.md [hunk 2]
```diff
@@ -1,3 +1,43 @@
...
+- new persistent gemm kernel
+- gemm benchmark tool using nearest neighbour approach.
+
+### benchmark commandline
+
+```
+ python gemm_benchmark.py
+```
+
+# streamk gemm script v0.2
```
Comment (neoblizz) [inline python/perf-kernels/streamk/README.md:14]: Is v0.2 relevant here? Or is this meant to be a changelog. If its not a changelog, and the features apply to both, consider moving the commands as a "Getting Started" or something for ease of use.

File: python/perf-kernels/streamk/gemm_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import torch
+import triton
+import numpy as np
+
+from utils.solution_selection import tunedtree, tunedarr, solution_params
+from utils.gemm_wrapper import matmul
+
+
+def selection_test():
+    sm = 304
```
Comment (neoblizz) [inline python/perf-kernels/streamk/gemm_benchmark.py]: Might be worth removing hard-coded sm counts:

```py
import torch

if torch.cuda.is_available():
    current_device_index = torch.cuda.current_device()
    current_device = torch.cuda.get_device_properties(current_device_index)
    num_sms = current_device.multi_processor_count
```

**Review by neoblizz — CHANGES_REQUESTED** [review]
Looks good to me, my biggest comment here is I wonder if there's a way to merge streamk_kernel and streamk_kernel_atomic with some constexpr, so it uses the same code but uses atomics for the partials instead. That seems to be the fundamental change between the two but a bunch of redundant code is being used.

File: python/perf-kernels/streamk/gemm_benchmark.py (hunk 3 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/gemm_benchmark.py]: done, thanks

File: python/perf-kernels/streamk/README.md (hunk 2 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/README.md:14]: it is working like change log

File: python/perf-kernels/streamk/utils/gemm_wrapper.py (hunk 1 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/streamk/utils/gemm_wrapper.py]: done, thanks

**Comment by xiaohuguo2023** [comment]
> Looks good to me, my biggest comment here is I wonder if there's a way to merge streamk_kernel and streamk_kernel_atomic with some constexpr, so it uses the same code but uses atomics for the partials instead. That seems to be the fundamental change between the two but a bunch of redundant code is being used.

the reason to put in atomics kernel here is for Ilia to test some accuracy issue of atomic_add, I would like to keep it separate for now just because we don't want to complicate the opt for our remainder part. We have next big task for reducing vgpr spills for the remainder.

**Review by neoblizz — APPROVED** [review]

**Comment by neoblizz** [comment]
> > Looks good to me, my biggest comment here is I wonder if there's a way to merge streamk_kernel and streamk_kernel_atomic with some constexpr, so it uses the same code but uses atomics for the partials instead. That seems to be the fundamental change between the two but a bunch of redundant code is being used.
> 
> the reason to put in atomics kernel here is for Ilia to test some accuracy issue of atomic_add, I would like to keep it separate for now just because we don't want to complicate the opt for our remainder part. We have next big task for reducing vgpr spills for the remainder.

Looks good, thanks!

**Review by zhanglx13 — APPROVED** [review]
